File: Project Euler/0065 - Convergents of e.cpp
#include <bits/stdc++.h>
using namespace std;
//Larger numbers, works in Python
int twoProd;
unsigned long gcd(unsigned long a, unsigned long b){
    if(a%b==0){
        return b;
    } else {
        return gcd(b,a%b);
    }
}
int getA(int n){
    int v =1;
    if(n%3==0){
        v = twoProd;
        twoProd-=2;
    }
    return v;
}
int main () {
    int n=50;
    twoProd = (n/3)*2;
    unsigned long num = 1;
    unsigned long den = getA(n);
    unsigned long divisor;
    n--;
    while (n>=2) {
        int ai = getA(n);
        unsigned long newDen = ai * den + num;
        num = den;
        den = newDen;
        divisor = gcd (num,den);
        num = num/divisor;
        den = den/divisor;
        n--;
    }
    num = num + den*2;
    divisor = gcd(num,den);
    num = num/divisor;
    cout<<num<<'/'<<den<<'\n';
    int sum = 0;
    while(num>0){
        sum += (num%10);
        num = num/10;
    }
    cout<<sum<<'\n';
    return 0;
}

//std::ios_base::sync_with_stdio(false);

File: Project Euler/0065 - Convergents of e.py
import math;
twoProd=0
def gcd(a,  b) :
    if(a%b == 0):
        return b
    else:
        return gcd(b,a%b)
    
def getA(n):
    global twoProd
    v=1
    if(n%3==0):
        v = twoProd
        twoProd -= 2
    return v
3

n=100
twoProd = int(n/3)*2
num = 1
den = getA(n)
divisor = 0
n = n -1;
while (n>=2):
    ai = getA(n)
    newDen = ai * den + num
    num = den
    den = newDen
    divisor = gcd (num,den)
    num = int(num/divisor)
    den = int(den/divisor)
    n = n -1;

num = num + den*2
divisor = gcd(num,den)
num = int(num/divisor)
print (num, '/', den,'\n')
sum = 0;
while(num>0):
    sum += (num%10);
    num = int(num/10);

print(sum);
